### ISSUE TITLE
Github shallow-since doesn't return one commit in the chain

### ISSUE THREAD

**Issue by Maxim-Mazurok** [isaacs/github#1804]
I have a fork: https://github.com/Maxim-Mazurok/DefinitelyTyped/commits/master
which is currently at https://github.com/Maxim-Mazurok/DefinitelyTyped/commit/c599db54c908b8e60d7a07896607830e6589b267

I'm doing this:
1. Clone only latest commit for each branch of my fork:
```sh
git clone --depth=1 https://github.com/Maxim-Mazurok/DefinitelyTyped --no-single-branch
```
2. Set upstream to original repo:
```sh
git remote add upstream https://github.com/DefinitelyTyped/DefinitelyTyped
```
3. Then, I'm getting my latest commit date:
```sh
git log -1 --format="%cd" --date=unix
```
4. Then I'm fetching commits from the main repo required to update my branch:
```sh
git fetch upstream master --shallow-since=1589324215
```

Then I'm trying to reset my branch by running:
```sh
git reset --hard upstream/master 
git push origin
```

But it results in reject:
```
$ git push origin
To https://github.com/Maxim-Mazurok/DefinitelyTyped
 ! [rejected]            master -> master (non-fast-forward)
error: failed to push some refs to 'https://github.com/Maxim-Mazurok/DefinitelyTyped'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

I did investigation, even when I run `git fetch upstream master --shallow-since=1589000000`
my local copy still doesn't have https://github.com/DefinitelyTyped/DefinitelyTyped/commit/9d282b0d7d2e05d7324e70afa66c4ff1306e67b7 commit.

I think that for some reason github just won't return it.

This results in:
```
$ git rev-list upstream/master --max-parents=0 HEAD
c7883f4581688909baf65c1219a8e842e369150b
9c96a1cd729d0209ba696f225b42d8aafd27295b
```

You can also see a number of failed "Auto Open PR" action runs as a result of this issue: https://github.com/Maxim-Mazurok/google-api-typings-generator/runs/670491014

Please, let me know if I'm doing something wrong here, or it's git / github bug.

Thanks!

(this issue was reported to GitHub and posted here for future references / ideas)

**Comment by Maxim-Mazurok** [comment]
Update:

**TL;DR** It's a got bug, reported here: https://lore.kernel.org/git/20200721160643.GA3288097@coredump.intra.peff.net/

> Jul 20, 2020, 12:36 PM UTC
> 
> Hi Maxim,
> 
> Thanks for writing in to GitHub Support and apologies for the substantial delay in getting back to you.
> 
> We really would have preferred to get back to you sooner. Unfortunately, it has been a while and I can see that your repository has moved on significantly since you last wrote in; could you confirm whether you are still experiencing similar issues?
> 
> If you have new details or examples of where you are running into this trouble; steps you've taken and any links or screenshots that would be super helpful so we can attempt to reproduce the issue and escalate this with the appropriate team.
> 
> Once we have more information, we'll certainly do our best to help!
> 
> All the best,
> Lesley
> GitHub Support
> 
> 


Thank you for getting back.
I still can reproduce the issue. Here are the steps to reproduce it regardless of how long it was since I sent this email to you:
1. cd /tmp
1. git clone --depth=1 https://github.com/Maxim-Mazurok/DefinitelyTyped --no-single-branch
1. cd DefinitelyTyped/
1. git remote add upstream https://github.com/DefinitelyTyped/DefinitelyTyped
1. git fetch upstream master --shallow-since=1589000000
1. git show 9d282b0d7d2e05d7324e70afa66c4ff1306e67b7
If you're getting error "fatal: bad object 9d282b0d7d2e05d7324e70afa66c4ff1306e67b7" - this means that issue is reproduced.

In order to get this commit, you can do all the same steps, but in the step 5, use 1588000000
In this case, commit gets fetched and you can run "git show -s --format=%ct 9d282b0d7d2e05d7324e70afa66c4ff1306e67b7" to get unix timestamp of that commit.
Which is 1589327434.
That means, that when I ask for "--shallow-since=1589000000" it should've been fetched.
Because commit date is 1589327434 = 05/12/2020
And I was asking for everything since 1589000000 = 05/09/2020
That's more than 3 days difference, so this issue can't be because of the timezones, I think.

Please, let me know if you can reproduce this or need any more input from me, I'll be happy to help!


> Hi Maxim,
> 
> Thank you for getting back to us and providing such detailed steps to reproduce the trouble you're experiencing.
> 
> I've asked our git experts for their input on this one. I can't promise an ETA for when we'll get an update, but we'll write back to you as soon as there's any news to share.
> 
> Please let us know if you have any other questions in the meantime.


> Hi Maxim,
> 
> Thank you for your patience while we looked into this.
> 
> Our engineers believe this could be an upstream problem with Git's shallow feature, unrelated to GitHub specifically.
> 
> Unfortunately, there isn't much we can do on our side to fix it. That said, our team have submitted a bug report to the Git mailing list. You're welcome to follow along here:
> 
> https://lore.kernel.org/git/20200721160643.GA3288097@coredump.intra.peff.net/
> 
> Sorry we couldn't be of more direct assistance this time.

Thank you, this bug report looks solid!
I was expecting this to be a git issue, bot a GitHub.
Thanks!
